[FLINK-31983] Add yarn Acls capability to Flink containers

architgyl opened the pull request:
## What is the purpose of the change
[FLINK-31983](https://issues.apache.org/jira/browse/FLINK-31983)
[FLIP-312](https://cwiki.apache.org/confluence/display/FLINK/FLIP+312%3A+Add+Yarn+ACLs+to+Flink+Containers)
This PR provides Yarn application Acl mechanism on Flink containers to be able to provide specific rights to users other than the one running the Flink application job. This will restrict other users in two ways:
- view logs through the Resource Manager job history
- kill the application

**Note**: This PR is based on an earlier PR and I have addressed all the comments on the previous PR. 
[GitHub Pull Request #8760](https://github.com/apache/flink/pull/8760)
## Brief change log
Two parameters have been added:
 - `yarn.view.acls`
 - `yarn.modify.acls`
 
Add these parameter in `flink-conf.yaml`.

## Verifying this change
Verified the changes manually by adding the parameters in the `flink-conf.yaml` file and running a WordCount job on a Yarn cluster. 
- Tested `yarn.view.acls` when application in Running state:
    - User submitting the job can view the logs.
    - Another user in the `yarn.view.acls` can view the logs.
    - Another user not in the list of `yarn.view.acls` cannot view the logs.
**Note**: Logs can be view from the ResourceManager (<resourcemanager_url>/cluster/app/<app_id>)
    
- Tested `yarn.modify.acls` when application in Running state:
    -  User submitting the job can kill the application job.
    - Another user in the `yarn.modify.acls` can kill the application job..
    - Another user not in the list of `yarn.modify.acls` cannot kill the application job.

- Tested with wildcard `*` giving access to everyone.
    
## Does this pull request potentially affect one of the following parts:

- Dependencies (does it add or upgrade a dependency): no
- The public API, i.e., is any changed class annotated with @Public(Evolving): no
- The serializers: no
- The runtime per-record code paths (performance sensitive): no
- Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, - ZooKeeper: Affect acls on YARN (currently not managed so no rights can be added)
- The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

venkata91 commented on .idea/vcs.xml:
```diff
@@ -3,6 +3,14 @@
       <list>
+        <IssueNavigationLink>
+          <option name="issueRegexp" value="[A-Z]+\-\d+" />
+          <option name="linkRegexp" value="https://go/jira $0" />
```
remove this file changes?

architgyl commented:
Tagging @becketqin @venkata91 for review.

venkata91 commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -231,6 +231,25 @@ public class YarnConfigOptions {
+            key("yarn.view.acls")
+                    .defaultValue("")
+                    .withDescription(
+                            "Users and groups to give VIEW acess. The ACLs are of for"
```
nit: typo in `access`. `acess` -> `access`?

venkata91 commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -231,6 +231,25 @@ public class YarnConfigOptions {
+            key("yarn.modify.acls")
+                    .defaultValue("")
+                    .withDescription(
+                            "Users and groups to give MODIFY acess. The ACLs are of for"
```
nit: typo in `access`. `acess` -> `access`?

venkata91 commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -425,6 +426,8 @@ static ContainerLaunchContext createTaskExecutorContext(
 
         ctx.setEnvironment(containerEnv);
 
+        setAclsFor(ctx, flinkConfig);
```
Is this to set `acls` to `TaskManager` containers and the `YarnClusterDescriptor` is for `AM` container?

mridulm replied:
This is within `createTaskExecutorContext` - so for TM.
The change in `YarnClusterDescriptor` is within `startAppMaster`, so for JM

venkata91 commented:
Tagging @GJL who reviewed the previous PR as well. Please review.

mridulm commented on flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNSessionFIFOITCase.java:
```diff
@@ -61,6 +61,9 @@
     private static final Logger log = LoggerFactory.getLogger(YARNSessionFIFOITCase.class);
 
+    protected static final String VIEW_ACLS = "user groupUser";
+    protected static final String MODIFY_ACLS = "admin groupAdmin";
```
Comma separated ?

architgyl replied:
[yarn_config_configuration.html](https://github.com/apache/flink/pull/22509/files#diff-4cc0177d60d77c8054982fe8c4b0edcd8e521d6a6afadd48bea95486cd236ae7) says usernames should be comma separated + <Space> + group names comma separated.
So, here `user` is showing username and `groupUser` is for Group name.

mridulm replied:
`groupUser` threw me off - `group` would be better :)

architgyl replied:
sure, changed.

mridulm commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -620,4 +623,30 @@ public static YarnConfiguration getYarnConfiguration(
+        }
+        if (modifyAcls != null) {
+            acls.put(ApplicationAccessType.MODIFY_APP, modifyAcls);
+        }
```
By default include `user.name` always ? (for view and modify acl)

architgyl replied:
User who is running the Flink application will always have the right to view the logs and kill the application irrespective of the `yarn.view.acls` and `yarn.modify.acls` values.

venkata91 replied:
Looks like Spark is explicitly adding `user.name and SPARK_USER` both to the acls by default. Although from this [doc](https://docs.cloudera.com/runtime/7.2.7/yarn-security/yarn-security.pdf), looks like the `Users who start an application (the owners) always have access to the application they start, which includes the application logs, job statistics, and ACLs`

venkata91 commented:
Gentle ping @GJL and @becketqin for review.

GJL commented:
Sorry, I don't have bandwidth at the moment and I don't remember the initial PR from 4 years ago. @rmetzger could you take a look?

rmetzger commented:
I'm going to be out of office for the next 2 weeks, and I haven't worked on the YARN code much recently. Maybe it would be easiest to look for folks who have recently worked on Flink on YARN.

venkata91 commented:
@gaborgsomogyi Could you help to review this PR?

venkata91 commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -620,4 +623,30 @@ public static YarnConfiguration getYarnConfiguration(
+            acls.put(ApplicationAccessType.MODIFY_APP, modifyAcls);
+        }
+        if (!acls.isEmpty()) {
+            amContainer.setApplicationACLs(acls);
```
What happens if wildcard (`*`) is passed in the config value? Should we only set the `*` instead of `user1,*`. Looks like YARN cannot recognize it properly based on the comment [here](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/SecurityManager.scala#L134)?

architgyl replied:
@venkata91 Tested with wildcard `*` and it works as expected which is giving all the users have access.

venkata91 replied:
No, try with `user1,*` and see what happens. We should handle it similar to spark. If there is `*` in the ACL, we should only set `*`.

architgyl replied:
Made the change and added special handling for wildcards.

venkata91 commented on docs/layouts/shortcodes/generated/yarn_config_configuration.html:
```diff
@@ -182,5 +182,15 @@
             <td>Specify YARN node label for the Flink TaskManagers, it will override the yarn.application.node-label for TaskManagers if both are set.</td>
         </tr>
+        <tr>
+          <td><h5>yarn.view.acls</h5></td>
```
Should we link to either this [doc](https://docs.cloudera.com/documentation/enterprise/latest/topics/cm_mc_yarn_acl.html) or this [doc](https://docs.cloudera.com/runtime/7.2.7/yarn-security/yarn-security.pdf)? Looks like there are more ACL rules and syntax. For eg:
1. Group names are case sensitive.
2. If an ACL starts with a single space, then it must consist of groups only.

architgyl replied:
@venkata91 Can you please suggest on the process to update the doc.

architgyl replied:
Thanks for sharing the other doc. I feel we should use the [doc](https://docs.cloudera.com/runtime/7.2.7/yarn-security/yarn-security.pdf). Please let me know your thoughts.

venkata91 commented on flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNSessionFIFOITCase.java:
```diff
@@ -61,6 +61,9 @@
     private static final Logger log = LoggerFactory.getLogger(YARNSessionFIFOITCase.class);
 
+    protected static final String VIEW_ACLS = "user group";
+    protected static final String MODIFY_ACLS = "admin groupAdmin";
```
Can we add a test for `wildcard` as well?

architgyl replied:
Added test case for wildcard.

becketqin commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -231,6 +231,25 @@ public class YarnConfigOptions {
+     */
+    public static final ConfigOption<String> APPLICATION_VIEW_ACLS =
+            key("yarn.view.acls")
+                    .defaultValue("")
```
We need to define the type first.
```
key("yarn.view.acls")
                    .stringType()
                    .defaultValue("")
```

architgyl replied:
Done.

becketqin commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -231,6 +231,25 @@ public class YarnConfigOptions {
+    /** Users and groups to give MODIFY access. */
+    public static final ConfigOption<String> APPLICATION_MODIFY_ACLS =
+            key("yarn.modify.acls")
+                    .defaultValue("")
```
dito above.

architgyl replied:
Done.

becketqin reviewed (commented):
Thanks for the patch. Left some comments.

Ideally we should have a test case verifying that the ACL actually works. Have you verified the ACL actually works with a real deployment in YARN?

architgyl commented on docs/layouts/shortcodes/generated/yarn_config_configuration.html:
```diff
@@ -182,5 +182,15 @@
+        <tr>
+          <td><h5>yarn.modify.acls</h5></td>
+          <td style="word-wrap: break-word;">(none)</td>
+          <td>Users and groups to give MODIFY access. The ACLs are of for comma-separated-usersspacecomma-separated-groups</td>
```
Updated the html file to display as :
```
Users and groups to give VIEW access. The ACLs are of for comma-separated-users<space>comma-separated-groups
Users and groups to give MODIFY access. The ACLs are of for comma-separated-users<space>comma-separated-groups
```

architgyl commented:
> Thanks for the patch. Left some comments.
> 
> Ideally we should have a test case verifying that the ACL actually works. Have you verified the ACL actually works with a real deployment in YARN?

@becketqin I have verified locally with the possible scenarios for both the use cases:
- view the logs
- kill application
Also, verified the wildcard behavior. Updated tests related to it and also, added post change behavior in the Testing part of the PR.

becketqin commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -620,4 +626,38 @@ public static YarnConfiguration getYarnConfiguration(
+        String viewAcls = flinkConfig.getString(YarnConfigOptions.APPLICATION_VIEW_ACLS, null);
+        String modifyAcls = flinkConfig.getString(YarnConfigOptions.APPLICATION_MODIFY_ACLS, null);
+        if (viewAcls != null) {
+            if (viewAcls.contains(WILDCARD_ACL)) {
```
Will this lead to unexpected behavior? For example, users may provide an acl like:
```
user1,user2,user_prefix_*
```
In this case, they might think the user_prefix_* are parse correctly as a regex, while we silently replaced the entire ACL with a `*`. It seems that if we do not support such case, an exception should be thrown.

architgyl replied:
@becketqin have updated the code with handling the wildcard use-case. Can you please review.

becketqin reviewed (commented):
Looks good to me overall. Just one comment. 

@wangyang0918 will you also have time to take a look? Thanks!

architgyl commented on flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNSessionFIFOITCase.java:
```diff
@@ -60,6 +60,12 @@
+    protected static final String VIEW_ACLS = "user group";
+    protected static final String MODIFY_ACLS = "admin groupAdmin";
+    protected static final String VIEW_ACLS_WITH_WILDCARD = "user,* group";
+    protected static final String MODIFY_ACLS_WITH_WILDCARD = "admin,* groupAdmin";
```
@becketqin addressed the comments. * makes sense rather than `user1,user2,*`. Can you please review.

becketqin commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -231,6 +231,27 @@ public class YarnConfigOptions {
+                    .defaultValue("")
+                    .withDescription(
+                            "Users and groups to give VIEW access. The ACLs are of for"
+                                    + " comma-separated-usersspacecomma-separated-groups");
```
We should modify the description here to:
```
Users and groups to give VIEW access. The ACLs are of for comma-separated-users&lt;space&gt;comma-separated-groups. Wildcard ACL is also supported. The only valid wildcard ACL is *, which grants permission to all the groups and users. 
```
The HTML files are generated from the description here.

Also, probably should mention the wildcard as well.

architgyl replied:
Fixed.

becketqin commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -231,6 +231,27 @@ public class YarnConfigOptions {
+                    .stringType()
+                    .defaultValue("")
+                    .withDescription(
+                            "Users and groups to give MODIFY access. The ACLs are of for"
```
Ditto above.

becketqin reviewed (approved):
LGTM. Just one minor comment.

wangyang0918 commented on .idea/vcs.xml:
```diff
@@ -1,25 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Why do we need to delete this file?

architgyl replied:
Restored the file

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -620,4 +626,46 @@ public static YarnConfiguration getYarnConfiguration(
+    public static void setAclsFor(
+            ContainerLaunchContext amContainer,
+            org.apache.flink.configuration.Configuration flinkConfig) {
+        Map<ApplicationAccessType, String> acls = new HashMap<>();
+        String viewAcls = flinkConfig.getString(YarnConfigOptions.APPLICATION_VIEW_ACLS, null);
+        String modifyAcls = flinkConfig.getString(YarnConfigOptions.APPLICATION_MODIFY_ACLS, null);
```
nit: These variables could be `final`.

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -231,6 +231,31 @@ public class YarnConfigOptions {
+    public static final ConfigOption<String> APPLICATION_VIEW_ACLS =
+            key("yarn.view.acls")
+                    .stringType()
+                    .defaultValue("")
```
I think we could use the `.noDefaultValue()` here

architgyl replied:
Fixed.

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -620,4 +626,46 @@ public static YarnConfiguration getYarnConfiguration(
+            ContainerLaunchContext amContainer,
+            org.apache.flink.configuration.Configuration flinkConfig) {
+        Map<ApplicationAccessType, String> acls = new HashMap<>();
+        String viewAcls = flinkConfig.getString(YarnConfigOptions.APPLICATION_VIEW_ACLS, null);
```
We could remove the explicit `null` after setting `.noDefaultValue()` for config option.

architgyl replied:
Fixed.

wangyang0918 reviewed (commented):
Thanks for working on improving the security of YARN integration. This PR already looks pretty good to me and I just left some minor comments.

Moreover, do the users need to enable kerberos when they want the `yarn.view.acls` and `yarn.modify.acls` to take effect?

architgyl commented:
> When I enabled the `kerberos` for YARN, the acl mechanism seems to work now. I not sure whether this is related.
My understanding is that when Kerberos is enabled, users need to authenticate themselves using their Kerberos credentials.

wangyang0918 commented:
> Based on the [documentation](https://docs.cloudera.com/documentation/enterprise/6/6.3/topics/cm_mc_yarn_acl.html#concept_activate_yarn_ACLs) here, it looks like yarn.acl.enable should be set to true to enable YARN ACLs. Just curious, whether it is set in your yarn-site.xml? Interesting, that it works after enabling Kerberos, is it because the other cluster or environment have this config set?

Yes. I already configured the `yarn.acl.enabled = true` and `yarn.admin.acl = hadoop` in YARN cluster.

Using knox to access the YARN UI makes the acl configurations taking effect. So it is not related with kerberos.

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -231,6 +231,31 @@ public class YarnConfigOptions {
+                    .withDescription(
+                            "Users and groups to give MODIFY access. The ACLs are of for"
+                                    + " comma-separated-users&lt;space&gt;comma-separated-groups."
+                                    + " Wildcard ACL is also supported. The only valid wildcard ACL "
```
```suggestion
                                    + " Wildcard ACL is also supported. The only valid wildcard ACL"
```
Redundant space.

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -231,6 +231,31 @@ public class YarnConfigOptions {
+                    .withDescription(
+                            "Users and groups to give VIEW access. The ACLs are of for"
+                                    + " comma-separated-users&lt;space&gt;comma-separated-groups."
+                                    + " Wildcard ACL is also supported. The only valid wildcard ACL "
```
```suggestion
                                    + " Wildcard ACL is also supported. The only valid wildcard ACL"
```
Redundant space.

wangyang0918 reviewed (changes requested):
Could you please regenerate the docs via `mvn package -Dgenerate-config-docs -pl flink-docs -am -nsu -DskipTests`? After then I think this PR is is ok to merge.

wangyang0918 commented:
I just want to squash the commit and push to `architgyl:master`. And I am not sure why this PR is accidentally closed.
